### PR TITLE
Weights & Biases Basic Integration

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes
This implements a basic version a wandb integration. It mimics the mlflow integration. I would like to get this version of the integration into the release for demo purposes only.

## TODO before release
- [x] Write proper README for example

## TODO Later after release
- [ ] Add stack component for wandb

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/wandb/wandb_step_decorator.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    _step: Optional[S] = None,
+    *,
+    project_name: Optional[str] = None,
+    experiment_name: Optional[str] = None,
+    entity: Optional[str] = None,
+) -> Union[S, Callable[[S], S]]:
+    """Decorator to enable wandb for a step function.
+
+    Apply this decorator to a ZenML pipeline step to enable wandb experiment
+    tracking. The wandb tracking configuration (tracking URI, experiment name,
```
Comment (schustmi) [inline src/zenml/integrations/wandb/wandb_step_decorator.py:64]: tracking uri, run name from mlflow

**Review by schustmi — APPROVED** [review]

File: examples/wandb_tracking/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+# Integrating wandb tracking into your pipeline
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:1]:
```suggestion
# Integrating Weights & Biases tracking into your pipeline
```

File: examples/wandb_tracking/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+[Weights&Biases](https://www.mlflow.org/docs/latest/tracking.html) is a popular
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:3]:
```suggestion
[Weights & Biases](https://www.mlflow.org/docs/latest/tracking.html) is a popular
```

File: examples/wandb_tracking/README.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+tool that tracks and visualizes experiment runs with their many parameters,
+metrics and output files.
+
+## Overview
+
+This example builds on the [quickstart](../quickstart) but showcases how easily
+Weights&Biases (wandb) tracking can be integrated into a ZenML pipeline.
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:10]:
```suggestion
Weights & Biases (`wandb`) tracking can be integrated into a ZenML pipeline.
```

File: examples/wandb_tracking/README.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+We'll be using the
+[MNIST](http://yann.lecun.com/exdb/mnist/) dataset and
+will train a classifier using [Tensorflow (Keras)](https://www.tensorflow.org/).
+We will run two experiments with different parameters (epochs and learning rate)
+and log these experiments into a wandb backend. 
+
+In the example script, we use the [Keras WandbCallback](https://docs.wandb.ai/ref/python/integrations/keras/wandbcallback) is
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:18]:
```suggestion
In the example script, the [Keras WandbCallback](https://docs.wandb.ai/ref/python/integrations/keras/wandbcallback) is
```

**Review by stefannica — APPROVED** [review]
I gave it a high-level review, I don't see any major issues.

File: examples/wandb_tracking/README.md (hunk 3 above)
Comment (AlexejPenner) [inline examples/wandb_tracking/README.md:3]: link to wandb not mlflow

File: examples/wandb_tracking/README.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+used within the training step to directly hook into the TensorFlow training and
+it will log out all relevant parameters, metrics and output files. Additionally,
+we explicitly log the test accuracy within the evaluation step.
+
+Note that despite wandb being used in different steps within a pipeline, ZenML handles initializing wandb
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:23]:
```suggestion
Note that despite `wandb `being used in different steps within a pipeline, ZenML handles initializing `wandb` 
```

File: examples/wandb_tracking/README.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+### Pre-requisites
+In order to run this example, you need to install and initialize ZenML:
+
+```shell
+# install CLI
+pip install zenml
+
+# install ZenML integrations
+zenml integration install tensorflow mlflow -f
```
Comment (AlexejPenner) [inline examples/wandb_tracking/README.md:37]: not mlflow

File: examples/wandb_tracking/README.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+and log these experiments into a wandb backend. 
+
+In the example script, we use the [Keras WandbCallback](https://docs.wandb.ai/ref/python/integrations/keras/wandbcallback) is
+used within the training step to directly hook into the TensorFlow training and
+it will log out all relevant parameters, metrics and output files. Additionally,
+we explicitly log the test accuracy within the evaluation step.
+
+Note that despite wandb being used in different steps within a pipeline, ZenML handles initializing wandb 
+and ensures the experiment name is the same as the pipeline name, and the experiment run is the same name 
+as the pipeline run name. This estabilishes a lineage between pipelines in ZenML and experiments in wandb.
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:25]:
```suggestion
as the pipeline run name. This establishes a lineage between pipelines in ZenML and experiments in `wandb`.
```

File: examples/wandb_tracking/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+# pull example
+zenml example pull wandb_tracking
+cd zenml_examples/wandb_tracking
+
+# initialize
+zenml init
+```
+
+### Set up Weights&Biases
+To get this example running, you need to set up a [Weights&Biases] account. You can do this for free [here](https://wandb.ai/login?signup=true).
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:48]:
```suggestion
To get this example running, you need to set up a [Weights & Biases] account. You can do this for free [here](https://wandb.ai/login?signup=true).
```

File: examples/wandb_tracking/README.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+After signing up, you will be given a username (what Weights&Biases calls an `entity`), and you can go ahead and create your first project.
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:50]:
```suggestion
After signing up, you will be given a username (what Weights & Biases calls an `entity`), and you can go ahead and create your first project.
```

**Review by AlexejPenner — CHANGES_REQUESTED** [review]

File: examples/wandb_tracking/README.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+Note, that in case you have a shared Weights&Biases account, the `entity` can also be your organization/teams name.
```
Comment (strickvl) [inline examples/wandb_tracking/README.md:52]:
```suggestion
Note, that in case you have a shared Weights & Biases account, the `entity` can also be your organization or team's name.
```

**Review by strickvl — CHANGES_REQUESTED** [review]
Some tiny nits. Otherwise looks good to me.

File: examples/wandb_tracking/README.md (hunk 3 above)
Comment (AlexejPenner) [inline examples/wandb_tracking/README.md]:
```suggestion
[Weights&Biases](https://wandb.ai/site/experiment-tracking) is a popular
```
